include/libdnf/system/state.hpp

```diff
@@ -0,0 +1,79 @@
+/*
+Copyright (C) 2021 Red Hat, Inc.
+
+This file is part of libdnf: https://github.com/rpm-software-management/libdnf/
+
+Libdnf is free software: you can redistribute it and/or modify
+it under the terms of the GNU Lesser General Public License as published by
+the Free Software Foundation, either version 2 of the License, or
+(at your option) any later version.
+
+Libdnf is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License
+along with libdnf.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#ifndef LIBDNF_SYSTEM_STATE_HPP
+#define LIBDNF_SYSTEM_STATE_HPP
+
+#include "libdnf/common/exception.hpp"
+#include "libdnf/transaction/transaction_item_reason.hpp"
+
+#include <filesystem>
+#include <string>
+#include <map>
+
+
+namespace libdnf::system {
+
+/// A class providing information and allowing modification of the DNF system
+/// state. The state consists of a list of userinstalled packages, installed
+/// groups and their packages etc.
+class State {
+public:
+    /// Creates an instance of `State`, optionally specifying the directory
+    /// where the state is stored.
+    /// @param dir_path The directory where the state is stored.
+    /// @since 5.0
+    State(const std::filesystem::path & installroot, const std::filesystem::path & dir_path = "/var/lib/dnf/state");
+
+    /// @return The reason for a package NA (Name.Arch).
+    /// @param na The NA to get the reason for.
+    /// @since 5.0
+    libdnf::transaction::TransactionItemReason get_reason(const std::string & na);
+
+    /// Sets the reason for a package NA (Name.Arch).
+    /// @param na The NA to set the reason for.
+    /// @param reason The reason to set.
+    /// @since 5.0
+    void set_reason(const std::string & na, libdnf::transaction::TransactionItemReason reason);
```

Comment:
I suggest that set_reason cannot be in public API. It will only create inconsistency between transaction a history DB and the system state.